File: python/164_Maximum_Gap.py
#!/usr/bin/env python3
import math

# 本机运行正确，在leetcode上运行不正确。可能是python版本问题
#


class ListNode(object):
        def __init__(self, val):
            self.val = val
            self.next = None

class Solution(object):
    def bucket_sort(self, array):
        """
        :param array: list[float] or list[(float, ...)] unsorted-array
        :return: list[float] or list[(float, ...)] sorted-array
        """
        n = len(array)
        linkArray = [ListNode(-1) for i in range(n)]
        result = []

        if isinstance(array[0], int): #根据是否附带卫星数据分情况
            for i in range(n):
                index = math.floor(n * array[i])
                newNode = ListNode(array[i])
                cur = linkArray[index]
                while cur.next:
                    if cur.next.val >= array[i]:
                        newNode.next = cur.next
                        cur.next = newNode
                        break
                    cur = cur.next
                else:
                    cur.next = newNode

            for i in range(len(linkArray)):
                if linkArray[i].next:
                    cur = linkArray[i].next
                    while cur:
                        result.append(cur.val)
                        cur = cur.next
        else:
            for i in range(n):
                index = int(math.floor(n * array[i][0]))
                newNode = ListNode(array[i])
                cur = linkArray[index]
                while cur.next:
                    if cur.next.val[0] >= array[i][0]:
                        newNode.next = cur.next
                        cur.next = newNode
                        break
                    cur = cur.next
                else:
                    cur.next = newNode

            for i in range(len(linkArray)):
                if linkArray[i].next:
                    cur = linkArray[i].next
                    while cur:
                        result.append(cur.val)
                        cur = cur.next

        return result

    def bucket_sort_pretreatment(self, array):
        new_array = []

        if isinstance(array[0], int):
            maxNum = max(array)
            minNum = min(array)
            length = maxNum - minNum + 1
            new_array = [((x-minNum)/length, x) for x in array]
        else:
            maxNum = max([x[0] for x in array])
            minNum = min([x[0] for x in array])
            length = maxNum - minNum + 1
            new_array = [((x[0]-minNum)/length, x) for x in array]

        return new_array

    def maximumGap(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if not nums or len(nums) < 2:
            return 0

        result = [x[1] for x in self.bucket_sort(self.bucket_sort_pretreatment(nums))]
        gap = 0
        for i in range(1, len(result)):
            if result[i] - result[i-1] > gap:
                gap = result[i] - result[i-1]

        return gap

s = Solution()
print(s.maximumGap([1,10000000]))

File: java/src/brand_new/_164_Maximum_Gap.java
package brand_new;

public class _164_Maximum_Gap {
	
	public int maximumGap(int[] nums) {
		if(nums == null || nums.length <= 1){
			return 0;
		}
		
		long leftEnd = Math.min(nums[0], nums[1]);
		long rightEnd = Math.max(nums[0], nums[1]);
		long maxGap = rightEnd - leftEnd;
		
		for(int i = 2; i < nums.length; i++){
			if(nums[i] > rightEnd){
				if(nums[i] - rightEnd > maxGap){
					maxGap = nums[i] - rightEnd;
				}
				rightEnd = nums[i];
			}else if(nums[i] < leftEnd){
				if(leftEnd - nums[i] > maxGap){
					maxGap = leftEnd - nums[i];
				}
				leftEnd = nums[i];
			}
		}

		return (int) maxGap;
	}

	public static void main(String[] args) {
		System.out.println(new _164_Maximum_Gap().maximumGap(new int[]{15252,16764,27963,7817,26155,20757,3478,22602,20404,6739,16790,10588,16521,6644,20880,15632,27078,25463,20124,15728,30042,16604,17223,4388,23646,32683,23688,12439,30630,3895,7926,22101,32406,21540,31799,3768,26679,21799,23740}));
	}
}
